File: src/frontend/folha/lib/app/data/model/folha_historico_salarial_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:folha/app/infra/infra_imports.dart';
import 'package:folha/app/data/model/model_imports.dart';

class FolhaHistoricoSalarialModel {
	int? id;
	int? idColaborador;
	String? competencia;
	double? salarioAtual;
	double? percentualAumento;
	double? salarioNovo;
	String? validoAPartir;
	String? motivo;
	ViewPessoaColaboradorModel? viewPessoaColaboradorModel;

	FolhaHistoricoSalarialModel({
		this.id,
		this.idColaborador,
		this.competencia,
		this.salarioAtual,
		this.percentualAumento,
		this.salarioNovo,
		this.validoAPartir,
		this.motivo,
		this.viewPessoaColaboradorModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'competencia',
		'salario_atual',
		'percentual_aumento',
		'salario_novo',
		'valido_a_partir',
		'motivo',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Competencia',
		'Salario Atual',
		'Percentual Aumento',
		'Salario Novo',
		'Valido A Partir',
		'Motivo',
	];

	FolhaHistoricoSalarialModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idColaborador = jsonData['idColaborador'];
		competencia = jsonData['competencia'];
		salarioAtual = jsonData['salarioAtual']?.toDouble();
		percentualAumento = jsonData['percentualAumento']?.toDouble();
		salarioNovo = jsonData['salarioNovo']?.toDouble();
		validoAPartir = jsonData['validoAPartir'];
		motivo = jsonData['motivo'];
		viewPessoaColaboradorModel = jsonData['viewPessoaColaboradorModel'] == null ? ViewPessoaColaboradorModel() : ViewPessoaColaboradorModel.fromJson(jsonData['viewPessoaColaboradorModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idColaborador'] = idColaborador != 0 ? idColaborador : null;
		jsonData['competencia'] = Util.removeMask(competencia);
		jsonData['salarioAtual'] = salarioAtual;
		jsonData['percentualAumento'] = percentualAumento;
		jsonData['salarioNovo'] = salarioNovo;
		jsonData['validoAPartir'] = Util.removeMask(validoAPartir);
		jsonData['motivo'] = motivo;
		jsonData['viewPessoaColaboradorModel'] = viewPessoaColaboradorModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idColaborador = plutoRow.cells['idColaborador']?.value;
		competencia = plutoRow.cells['competencia']?.value;
		salarioAtual = plutoRow.cells['salarioAtual']?.value?.toDouble();
		percentualAumento = plutoRow.cells['percentualAumento']?.value?.toDouble();
		salarioNovo = plutoRow.cells['salarioNovo']?.value?.toDouble();
		validoAPartir = plutoRow.cells['validoAPartir']?.value;
		motivo = plutoRow.cells['motivo']?.value;
		viewPessoaColaboradorModel = ViewPessoaColaboradorModel();
		viewPessoaColaboradorModel?.nome = plutoRow.cells['viewPessoaColaboradorModel']?.value;
	}	

	FolhaHistoricoSalarialModel clone() {
		return FolhaHistoricoSalarialModel(
			id: id,
			idColaborador: idColaborador,
			competencia: competencia,
			salarioAtual: salarioAtual,
			percentualAumento: percentualAumento,
			salarioNovo: salarioNovo,
			validoAPartir: validoAPartir,
			motivo: motivo,
		);			
	}

	
}

File: backend/folha/src/model/folha_historico_salarial_model.py
from src import db
from src.model.view_pessoa_colaborador_model import ViewPessoaColaboradorModel


class FolhaHistoricoSalarialModel(db.Model):
    __tablename__ = 'folha_historico_salarial'

    id = db.Column(db.Integer, primary_key=True)
    competencia = db.Column(db.String(7))
    salario_atual = db.Column(db.Float)
    percentual_aumento = db.Column(db.Float)
    salario_novo = db.Column(db.Float)
    valido_a_partir = db.Column(db.String(7))
    motivo = db.Column(db.String(100))
    id_colaborador = db.Column(db.Integer, db.ForeignKey('view_pessoa_colaborador.id'))

    view_pessoa_colaborador_model = db.relationship('ViewPessoaColaboradorModel', foreign_keys=[id_colaborador])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_colaborador = data.get('idColaborador')
        self.competencia = data.get('competencia')
        self.salario_atual = data.get('salarioAtual')
        self.percentual_aumento = data.get('percentualAumento')
        self.salario_novo = data.get('salarioNovo')
        self.valido_a_partir = data.get('validoAPartir')
        self.motivo = data.get('motivo')

    def serialize(self):
        return {
            'id': self.id,
            'idColaborador': self.id_colaborador,
            'competencia': self.competencia,
            'salarioAtual': self.salario_atual,
            'percentualAumento': self.percentual_aumento,
            'salarioNovo': self.salario_novo,
            'validoAPartir': self.valido_a_partir,
            'motivo': self.motivo,
            'viewPessoaColaboradorModel': self.view_pessoa_colaborador_model.serialize() if self.view_pessoa_colaborador_model else None,
        }